lib/src/main/java/com/knuddels/jtokkit/Cl100kParser.java

```diff
@@ -0,0 +1,247 @@
+                    || ((lch1 == 'V') && (lch2 == 'E'))
+                    || ((lch1 == 'R') && (lch2 == 'E'));
+        }
+    }
+
+    public static boolean isValidUTF8(String input) {
+        return UTF_8.newEncoder().canEncode(input);
+    }
+
+    static boolean isLetter(int ch) {
+        if (ch < 0xaa) {
+            return ((ch >= 'a') && (ch <= 'z'))
+                    || ((ch >= 'A') && (ch <= 'Z'));
+        } else if (ch <= 0x323af) {
+            switch (getType(ch)) {
+                case UPPERCASE_LETTER:
+                case LOWERCASE_LETTER:
+                case TITLECASE_LETTER:
+                case MODIFIER_LETTER:
+                case OTHER_LETTER:
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    static boolean isNumeric(int ch) {
+        if (ch < 0xb2) {
+            return (ch >= '0') && (ch <= '9');
+        } else if (ch <= 0x1fbf9) {
+            switch (getType(ch)) {
+                case DECIMAL_DIGIT_NUMBER:
+                case LETTER_NUMBER:
+                case OTHER_NUMBER:
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    static boolean isLetterOrNumeric(int ch) {
+        if (ch < 0xaa) {
+            return ((ch >= 'a') && (ch <= 'z'))
+                    || ((ch >= 'A') && (ch <= 'Z'))
+                    || ((ch >= '0') && (ch <= '9'));
+        } else if (ch <= 0x323af) {
+            switch (getType(ch)) {
+                case UPPERCASE_LETTER:
+                case LOWERCASE_LETTER:
+                case TITLECASE_LETTER:
+                case MODIFIER_LETTER:
+                case OTHER_LETTER:
+                case DECIMAL_DIGIT_NUMBER:
+                case LETTER_NUMBER:
+                case OTHER_NUMBER:
+                    return true;
+            }
+        }
+        return false;
+    }
+
+    static boolean isWhitespace(int ch) {
+        if (ch <= '\r') {
+            return SIMPLE_WHITESPACES.indexOf(ch) >= 0;
+        } else if (ch < '\u0085') {
+            return ch == ' ';
+        } else {
+            return (ch == '\u0085')
+                    || (ch == '\u00A0')
+                    || ((ch >= '\u1680') && (ch <= '\u3000') && (binarySearch(REMAINING_WHITESPACES, ch) >= 0));
+        }
+    }
+
+    static boolean isNewline(int ch) {
+        return (ch == '\r')
+                || (ch == '\n');
+    }
+
+    static boolean isNotWhitespaceOrLetterOrNumeric(int ch) {
+        if (ch < '0') {
+            return ch >= 0 && ch != ' ' && (ch > '\r' || ch < '\t');
+        } else {
+            return !isLetterOrNumeric(ch) && !isWhitespace(ch);
+        }
+    }
+
+    static boolean isNotNewlineOrLetterOrNumeric(int ch) {
+        if (ch < '0') {
+            return ch >= 0 && (ch == ' ' || !isNewline(ch));
+        } else {
+            return !isLetterOrNumeric(ch);
+        }
+    }
+
+    static List<Byte> addUtf8Bytes(String input, int start, int end, List<Byte> dst) {
```

Comment:
instead of getting the UTF8 bytes from the string to a new byte[], we're converting it to a reusable list